webhooks/services.py

```diff
@@ -1,14 +1,137 @@
+        return to_proof_hook_model(item=item)
+
+    def _credential_hook_versioned(self, item: dict) -> dict:
+        return to_credentential_hook_model(item=item)
+
+    def _version_connections(self, item: dict) -> dict:
+        return to_connections_model(item=item)
+
+    def _deserialise(self, data: bytes) -> Union[str, dict]:
+        # Turn bytes from redis into str
+        try:
+            return (
+                data.decode("utf-8")
+                .replace("'", '"')
+                .replace("False", json.dumps(False))
+                .replace("True", json.dumps(True))
+                .replace("None", json.dumps(None))
+            )
+        except JSONDecodeError:
+            # Don't raise an error here. It's probably more valuable to
+            # Return the data anyway than breaking the retrieval
+            return {"invalid JSON data": data}
+
+    def _to_json(self, data: List[bytes]) -> List[dict]:
+        data = [self._deserialise(val) for val in data]
+        try:
+            data = [json.loads(d) for d in data]
+        except JSONDecodeError:
+            # A multiline string has occured that has been split in
+            # an invalid way wrt JSON format
+
+            # Remove multiple spaces which occur when error msg is split into multiple lines
+            # turn off black formatter here because it replaces the regex
+            # fmt: off
+            data = [re.sub('\s+',' ', d) for d in data]
+            # Remove [..\" \"..] from the multiline error message because that is invalid json
+            data = [d.replace('\" \"', ' ') for d in data]
+            # fmt: on
+            data = [json.loads(d) for d in data]
+        return data
+
+    def _to_item(self, topic: str, data: List[dict]) -> List[dict]:
+        # Transform the data to the appropriate model
+        data = self._to_json(data=data)
+        if self._is_proof(topic):
+            data = [self._proof_hook_versioned(d) for d in data]
+        elif self._is_credential(topic):
+            data = [self._credential_hook_versioned(d) for d in data]
+        elif topic == "connections":
+            data = [self._version_connections(d) for d in data]
+        elif topic == "basicmessages":
+            data = [BasicMessagesHook(**d) for d in data]
+        return data
+
+    def _to_topic_item(self, topic: str, data: List[bytes]) -> List[TopicItem]:
+        data = self._to_item(topic=topic, data=data)
+        return (
+            [
+                TopicItem(
+                    topic=topic,
+                    wallet_id=payload.wallet_id,
+                    origin=payload.origin,
+                    payload=payload.dict(),
+                )
+                for payload in data
+            ]
+            if data
+            else []
+        )
+
+    async def add_topic_entry(self, topic: str, hook: bytes) -> List[TopicItem]:
         return await self._redis.sadd(topic, hook)
```

Comment:
Couldn't we transform the hook to json here?

Reply:
In theory we can. And I can have a go at it. However storing JSON in redis works a bit different and we'll then have to always provide a unique non-existing key to the hook that is in json. This way I could just dump it all. Happy to change it/give it a try though

Reply:
Can redis differentiate between a json object encoded as string encoded as bytes and a dict encoded as string? I'm not talking about JSON directly, but rather a JSON string

Reply:
not sure I understand your question correctly, but have a look at [this](https://developer.redis.com/howtos/redisjson/using-python/)

Reply:
Just a quick thought you've probably suggested the better way of doing this below either way. If we store by wallet_id this gets much simpler. Only thing then is what if we dont have a wallet id (or do we always have a wallet_id)?